Mathlib/Topology/Algebra/IsOpenUnits.lean

```diff
@@ -0,0 +1,90 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.Jacobson.Ideal
+import Mathlib.Topology.Algebra.GroupWithZero
+import Mathlib.Topology.Algebra.Nonarchimedean.AdicTopology
+
+/-!
+
+# Topological monoids with open units
+
+We say that a topological monoid `M` has open units (`IsOpenUnits`) if `Mˣ` is open in `M` and
+has the subspace topology (i.e. inverse is continuous).
+
+Typical examples include monoids with discrete topology, topological groups (or fields),
+and rings `R` equipped with the `I`-adic topology where `I ≤ J(R)` (`IsOpenUnits.of_isAdic`).
+
+This condition is needed so that for an affine scheme `X` over `R` and `U ⊆ X` an affine subscheme,
+`U(R)` is an open subspace of `X(R)`.
+-/
+
+open Topology
+
+/--
+We say that a topological monoid `M` has open units if `Mˣ` is open in `M` and
+has the subspace topology (i.e. inverse is continuous).
+
+Typical examples include monoids with discrete topology, topological groups (or fields),
+and rings `R` equipped with the `I`-adic topology where `I ≤ J(R)`.
```

Comment:
What are typical non-examples? More precisely: in what setting do you want to use this, and why does it not follow from the rest of your context?

Reply:
A typical non-example is $$\mathbb{A}_K$$ because the topology on ideles is not the induced topology from adeles.

> in what setting do you want to use this

This condition is needed so that for an affine scheme `X` over `R` and `U ⊆ X` an affine subscheme,
`U(R)` is an open subspace of `X(R)`. (and is equivalent to it because $\mathbb{G}_m$ is an affine open subscheme of $\mathbb{G}_a$)

> why does it not follow from the rest of your context

I'm not sure what you are referring to by "the rest of your context", but I don't think there is an alternative typeclass general enough to cover both topological fields and adic local rings. Plus this is probably the right condition to consider given the equivalence to a well-behaved topology on $$R$$-points.

Reply:
Thanks for those explanations. Could you please incorporate them in the (module) docstring?